### ISSUE TITLE
How to get multi-gpu to work?

### ISSUE THREAD

**Issue by hsm207** [batra-mlp-lab/visdial-challenge-starter-pytorch#8]
I created an Compute Engine instance on Google Cloud Compute with 4 K80 gpus, followed the instructions on the repo to setup the Anaconda environment and download the data. I ran the training with:

`python train.py --gpu-ids 0 1 2 3 `

The `batch_size` is 128 and `cpu_workers` is 4.

During training, I use `nvidia-smi` and can see that all 4 gpus are utilized (but rarely at 100%). Furthermore, the the seconds per iteration is a lot worse compared to a single GPU (8 vs 2).

What other configs should I adjust to get a speedup from using mulitple gpus?

**Comment by abhshkdz** [comment]
8 seconds per iteration sounds high. We were getting close to ~3 seconds per iteration on 4 GPUs and ~1 second per iteration on 1 GPU. 8 vs. 2 means they'll take roughly the same time to converge.

You might have to tweak the hyperparameters a bit if you're running on a single GPU (reduce the base learning rate for example).

You could consider increasing the number of workers to see if that gets you some speedup. Adding breakpoints in `getitem` and `train.py` to measure execution times might also provide hints on what the bottleneck is.

**Comment by hsm207** [comment]
@abhshkdz would you mind sharing the details of the machine you ran the training script on (what GPUs, cloud provider, etc)?

I doubled `cpu_worders` to 8 and it did not make a difference.

**Comment by abhshkdz** [comment]
These numbers are from Titan Xps using pytorch v1.0.0 built with cuda 9 on our cluster at Georgia Tech.  

> Also, if you are getting ~1 second per iteration on 1 GPU, should't you be getting less than ~1 second per iteration when using multiple GPUs?

The multi-GPU version does use a larger batch size, and there might be some overhead in syncs, not sure.

On n GPUs, we should be able to fit n times the single-GPU batch size, so it should take (roughly) 1/n times the no. of iterations as that of a single GPU (assuming hyperparameters tweaked accordingly).

So as long each iteration takes < nT (where T is time for one iteration on a single GPU), there should be some gains in overall convergence time.

**Comment by hsm207** [comment]
Thanks for the clarification.
